src/main/java/server/RequestParser.java

```diff
@@ -0,0 +1,72 @@
+package server;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.util.HashMap;
+
+public class RequestParser {
+    private BufferedReader in;
+
+    public RequestParser(BufferedReader in) {
+        this.in = in;
+    }
+
+    public HashMap parse() {
+        HashMap parsedRequest = new HashMap();
+
+        parsedRequest.put("requestLine", parseRequestLine());
+
+        HashMap<String, String> requestHeaders = parseRequestHeaders();
+        parsedRequest.put("requestHeaders", requestHeaders);
+
+        Integer contentLength = parseRequestBodyContentLength(requestHeaders);
+        parsedRequest.put("requestBody", parseRequestBody(contentLength));
+
+        return parsedRequest;
+    }
+
+    private String parseRequestLine() {
+        String requestLine = null;
+        try {
+            requestLine = in.readLine();
+        } catch (IOException err) {
+            System.err.println(err);
+        }
+        return requestLine;
+    }
+
+    private HashMap parseRequestHeaders() {
+        HashMap<String, String> requestHeaders = new HashMap<>();
+        try {
+            String header = in.readLine();
+            while (header != null && !header.isEmpty()) {
+                String[] headerComponents = header.split(": ");
+                requestHeaders.put(headerComponents[0].trim(), headerComponents[1].trim());
+                header = in.readLine();
+            }
+        } catch (IOException err) {
+            System.err.println(err);
+        }
+        return requestHeaders;
+    }
+
+    private String parseRequestBody(int contentLength) {
+        StringBuilder requestBody = new StringBuilder();
+        int nextBodyChar;
+        try {
+            for (int i = 0; i < contentLength; i++) {
+                nextBodyChar = in.read();
+                requestBody.append((char) nextBodyChar);
```

Comment:
this for-loop seems really inefficient. I believe you should be able to achieve the same thing with this and a char array: https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html#read-char:A-int-int-

Reply:
Oh that's much cleaner. Thanks!